File: 1319-number-of-operations-to-make-network-connected/1319-number-of-operations-to-make-network-connected.js
/**
 * @param {number} n
 * @param {number[][]} connections
 * @return {number}
 */
var makeConnected = function(n, connections) {
  const uf = new UnionFind(n);
  let extract = 0;
  
  for (let [a, b] of connections) {
    if (uf.connected(a, b)) extract++;
    else uf.union(a, b);
  }
  
  return extract >= uf.getCount() - 1 ? uf.getCount() - 1 : -1;
};

class UnionFind {
  constructor(n) {
    this.parent = Array.from({ length: n }, (_, i) => i);
    this.count = n;
  }

  find(x) {
    const parent = this.parent;
    while (x !== parent[x]) {
      parent[x] = parent[parent[x]];
      x = parent[x];
    }

    return x;
  }
  
  connected(x, y) {
    return this.find(x) === this.find(y);
  }

  getCount() {
    return this.count;
  }

  union(x, y) {
    const rootX = this.find(x);
    const rootY = this.find(y);
    if (rootX === rootY) return;

    rootX < rootY
      ? this.parent[rootX] = rootY
      : this.parent[rootY] = rootX;
    
    this.count--;
  }
}

File: 1319-number-of-operations-to-make-network-connected/1319-number-of-operations-to-make-network-connected.py
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.count = n
        
    def find(self, x):
        if x != self.parent[x]:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x] 
      
    def componentSize(self):
        return self.count
      
    def union(self, x, y):
        root_x, root_y = self.find(x), self.find(y)
        if root_x == root_y:
            return
        self.parent[root_x] = root_y
        self.count -= 1
        

class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        if len(connections) < n - 1:
            return -1
        
        uf = UnionFind(n)
        for a, b in connections:
            uf.union(a, b)
        return uf.componentSize() - 1